### PR TITLE
proxy: improve ABCI app connection handling

### LINKED ISSUE #5074: Tendermint does not exit when ABCI app stops and
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):
0.33.5-e22acc0f

**ABCI app** (name for built-in, URL for self-written if it's publicly available):
abci-cli kvstore

**Environment**:
- **OS** (e.g. from /etc/os-release):
```
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://www.archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
LOGO=archlinux
```
- **Install tools**: N/A
- **Others**: N/A

**What happened**:
Tendermint does not terminate when the ABCI app stops when running with `--consensus.create_empty_blocks=false`

**What you expected to happen**:
I would expect tendermint to exit as described here: https://docs.tendermint.com/master/tendermint-core/running-in-production.html#what-happens-when-my-app-dies

It does however exit when you run tendermint without disabling the empty block creation

**Have you tried the latest version**: yes

**How to reproduce it** (as minimally and precisely as possible):

```bash
# terminal 1
tendermint init
tendermint node --consensus.create_empty_blocks=false
# terminal 2
abci-cli kvstore
# after connect, hit ctrl-c on abci-cli
```

**Logs (paste a small part showing an error (< 10 lines) or link a pastebin, gist, etc. containing more of the log file)**:

```bash
❯ tendermint node --consensus.create_empty_blocks=false
I[2020-07-01|20:28:13.664] Version info                                 module=main software=0.33.5 block=10 p2p=7
I[2020-07-01|20:28:13.672] Starting Node service                        module=main impl=Node
I[2020-07-01|20:28:13.681] Started node                                 module=main nodeInfo="{ProtocolVersion:{P2P:7 Block:10 App:1} DefaultNodeID:e5a6ca09229041270896b1598faee0d0d56859b5 ListenAddr:tcp://0.0.0.0:26656 Network:test-chain-g6Kt1w Version:0.33.5 Channels:4020212223303800 Moniker:engel Other:{TxIndex:on RPCAddress:tcp://127.0.0.1:26657}}"
I[2020-07-01|20:28:14.706] Executed block                               module=state height=1 validTxs=0 invalidTxs=0
I[2020-07-01|20:28:14.709] Committed state                              module=state height=1 txs=0 appHash=0000000000000000
I[2020-07-01|20:28:15.730] Executed block                               module=state height=2 validTxs=0 invalidTxs=0
I[2020-07-01|20:28:15.732] Committed state                              module=state height=2 txs=0 appHash=0000000000000000
E[2020-07-01|20:29:05.117] Stopping abci.socketClient for error: EOF    module=abci-client connection=mempool
E[2020-07-01|20:29:05.118] Stopping abci.socketClient for error: EOF    module=abci-client connection=query
E[2020-07-01|20:29:05.118] Stopping abci.socketClient for error: EOF    module=abci-client connection=consensus
```
It displays the errors, but does not quit.

When running with empty block creation true, it does display the normal behavior. See [gist with log](https://gist.github.com/stijnopheide/ac4c0c20bd6ed3d0400a23600337e6e7).

**Config (you can paste only the changes you've made)**: none

**node command runtime flags**: `--consensus.create_empty_blocks=false`

**`/dump_consensus_state` output for consensus bugs**

**Anything else we need to know**: Not sure if this is related to https://github.com/tendermint/tendermint/issues/3388, but I cannot reproduce the behavior of 3388 without setting empty block creation to false.

### PR DESCRIPTION
**Pull request by melekes**
Closes #5074 

Old code does not work when --consensus.create_empty_blocks=false
(because it only calls tmos.Kill when ApplyBlock fails). New code is
listening ABCI clients for Quit and kills TM process if there were any
errors.

### CODE REVIEW

**Comment by melekes** [comment]
> are there any tests for this?

added

File: proxy/multi_app_conn_test.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+	appConns.Start()
+	clientMock.AssertCalled(t, "Start")
+
+	appConns.Stop()
+	clientMock.AssertCalled(t, "Stop")
+}
+
+// Upon failure, we call tmos.Kill
+func TestAppConns_Failure(t *testing.T) {
```
Comment (melekes) [inline proxy/multi_app_conn_test.go:40]: not sure how useful these tests are actually (cc @marbar3778 )

Comment (marbar3778) [reply proxy/multi_app_conn_test.go:40]: some tests are better than no tests 😄 they seem simple enough that they may not be useful but helps to have them just incase. 

it doesnt hurt to have them right?

File: proxy/multi_app_conn.go [hunk 2]
```diff
@@ -48,70 +59,118 @@ func NewMultiAppConn(clientCreator ClientCreator) AppConns {
...
 
-// Returns the snapshot Connection
 func (app *multiAppConn) Snapshot() AppConnSnapshot {
 	return app.snapshotConn
 }
 
 func (app *multiAppConn) OnStart() error {
-	// query connection
-	querycli, err := app.clientCreator.NewABCIClient()
+	c, err := app.abciClientFor("query")
```
Comment (erikgrinaker) [inline proxy/multi_app_conn.go:79]: We should add constants for these.

File: proxy/multi_app_conn_test.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	clientMock.On("Start").Return(nil)
+	clientMock.On("Stop").Return(nil)
+	clientMock.On("Quit").Return(quitCh)
+
+	clientCreatorMock.On("NewABCIClient").Return(clientMock, nil).Times(4)
+
+	appConns := NewAppConns(clientCreatorMock)
+
+	appConns.Start()
+	clientMock.AssertCalled(t, "Start")
```
Comment (erikgrinaker) [inline proxy/multi_app_conn_test.go:33]: We probably want to check that all connections have been started and stopped. One way to do this is to add `.Times(4)` onto the `On()` calls above, and then check `clientMock.AssertExpectations()` which asserts that all defined expectations have in fact been called the correct number of times (if given).

File: proxy/multi_app_conn_test.go (hunk 1 above)
Comment (melekes) [reply proxy/multi_app_conn_test.go:40]: I guess, yeah

**Pull request merged**